5653 - "Upgrade" dialog does not appear when upgrading from 3.4 > 3.5

kennsippell opened the pull request:
# Description

> Upgrading from master to this change doesn't give the user the "Update" prompt because there is no change to a document being watched in v3.4. Since this change won't be deployed atomically and will be instead packaged with other changes in v3.5 - I think that is safe. But we can consider tombstoning the legacy appcache document to make this change prompt an upgrade and also keep the database a bit cleaner. Looking for advice as I'm unsure how to best handle that.

https://github.com/medic/medic/issues/5653#issuecomment-491642151

Turns out that modifying the `appcache` doc won't work to trigger the upgrade since the doc's name has been removed from the `validate_doc_update` and `docs_by_replication_key` etc. So I don't replicate down anymore once the 3.5 service stack is running.

Instead, this proposed changes introduces a benign attribute into the medic_client ddoc to trigger the update. If this approach seems non-hacky, we can merge this into the branch for #5648 and AT that single PR.

medic/medic#5653

# Code review items

- Readable: Concise, well named, follows the [style guide](https://github.com/medic/medic-docs/blob/master/development/style-guide.md), documented if necessary.
- Documented: Configuration and user documentation on [medic-docs](https://github.com/medic/medic-docs/)
- Tested: Unit and/or e2e where appropriate
- Internationalised: All user facing text
- Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in Changes.md.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

kennsippell commented:
Or we could even just wait for https://github.com/medic/medic/pull/5640 to merge - since it contains legit changes in medic-client.

kennsippell commented:
@newtewt I was able to view the update via the following steps:

1. Deploy a previous version via docker + horti (I tried 3.2.1 and 3.4.0)
1. View the page via a valid https cert. Either with docker certs or ngrok.
1. Log in as an offline user.
1. Deploy branch `5646-update-menu` via horti (3.5.x)
1. Either wait ~5 minutes or click "Sync Now"
1. You should see an Update Now button appear.

Can I ask if the steps you were using are different in any way? Or if you can describe in further detail what you saw?

newtewt commented:
@kennsippell  I'm doing those same steps.  I see in the logs details about replication and appcache. The version in the about page is updated to latest but no prompt. Let's try and get some time together today?

Console logs below. 

```
Replication started after 300.04 seconds since previous attempt.
/medic/_design/medic/_rewrite/#/messages/::1 Application Cache Checking event
/medic/_design/medic/_rewrite/#/messages/::1 Application Cache Obsolete event
inbox.js:3 Replication ended after 3.093 seconds with status not_required
inbox.js:3 n {status: 404, name: "not_found", message: "missing", error: true, reason: "missing", …} "Possibly unhandled rejection: {"status":404,"name":"not_found","message":"missing","error":true,"reason":"missing","docId":"partners"}"
(anonymous) @ inbox.js:3
(anonymous) @ inbox.js:3
c @ inbox.js:3
$digest @ inbox.js:3
(anonymous) @ inbox.js:3
completeTask @ inbox.js:3
(anonymous) @ inbox.js:3
setTimeout (async)
o.defer @ inbox.js:3
$evalAsync @ inbox.js:3
(anonymous) @ inbox.js:3
l @ inbox.js:3
t @ inbox.js:3
d @ inbox.js:3
(anonymous) @ inbox.js:3
(anonymous) @ inbox.js:3
(anonymous) @ inbox.js:3
x @ inbox.js:3
_resolve @ inbox.js:3
(anonymous) @ inbox.js:3
(anonymous) @ inbox.js:3
requestAnimationFrame (async)
o @ inbox.js:3
n @ inbox.js:3
(anonymous) @ inbox.js:3
o @ inbox.js:3
push @ inbox.js:3
enter @ inbox.js:3
enter @ inbox.js:3
(anonymous) @ inbox.js:3
(anonymous) @ inbox.js:3
(anonymous) @ inbox.js:3
S @ inbox.js:3
configUpdated @ inbox.js:3
(anonymous) @ inbox.js:3
t.sync @ inbox.js:3
a @ inbox.js:3
t.invokeHook @ inbox.js:3
(anonymous) @ inbox.js:3
t.runAllHooks @ inbox.js:3
(anonymous) @ inbox.js:3
(anonymous) @ inbox.js:3
(anonymous) @ inbox.js:3
$digest @ inbox.js:3
$apply @ inbox.js:3
(anonymous) @ inbox.js:3
completeTask @ inbox.js:3
(anonymous) @ inbox.js:3
setTimeout (async)
o.defer @ inbox.js:3
a @ inbox.js:3
(anonymous) @ inbox.js:3
dispatch @ inbox.js:3
v.handle @ inbox.js:3
```

kennsippell commented:
The code which triggers the update dialog requires a change in the application cache or a failure during the update.
https://github.com/medic/medic/blob/3.4.x/webapp/src/js/controllers/inbox.js#L772

This was hard to reproduce because two changing results in the update dialog. The first clears the cache, and the second fails with error "no appcache to update." Unsure why some horti scenarios were triggering two doc changes while others were changing only one.

My proposed fix is to serve an empty appcache file to vPrevious. This triggers the update on the first document change.

kennsippell commented:
To upgrade to 3.5, we have the following requirements:

1. Clear the appcache's "master" record for the route `/medic/_design/medic/_rewrire/` or else the page will just load from cache and not get the new content
1. Trigger an "update now" prompt

In previous versions of the app, we attempt to update the appcache whenever one of the following docs change: `_design/medic`, `_design/medic-client`, `appcache`, or `settings`. We show the "Update Ready" dialog on any of these results occur:

1. The manifest undergoes no change
1. The manifest changes, and the resulting update succeeds.
1. The update throws an exception

### 1. Manifest undergoes no change
The current manifest caches `/medic/_design/medic/_rewire/`. If we change nothing, the upgrade is blocked by this cache entry. So we have to do something.

### 2. Manifest changes, update succeeds
"An application cache always includes at least one resource..." [[source](https://developer.mozilla.org/en-US/docs/Web/HTML/Using_the_application_cache#Resources_in_an_application_cache)] due to its master entry (it caches the page that requests the cache). So we serve up an empty manifest, it still caches `/medic/_design/medic/_rewire/` and the update is blocked.

### 3. The update throws an exception
We will get an update if we get an exception. For appcache: exceptions are different from errors. Errors occur if the manifest is invalid or the manifest includes a resource which isn't available.

My current approach for the update uses this method. There is a change to `_design/medic-client` doc, which triggers an appcache update. The request for the manifest yields a 404. The appcache moves to an "obsolete" state (no update triggered). A second change to the docs listed above (_design/medic-client or settings) triggers a second update which causes an exception "cannot update appcache since none exists". 


## Options
Maybe there is something tricky we can do with the appcache manifest.. but I'm running out of ideas there.

I think option 3 from above is our best option. But we need two document changes to get the dialog... So a change in `_design/medic-client` but then we could do one of ...

1. Cause a change to the `settings` doc. Potentially with `medic-conf` step during post-deployment?
1. Cause a change to the `appcache` couchdb doc. I removed the `appcache` doc from couchdb and it no longer replicates down to the device as of 3.5. But we could undo that part of the service-worker change and replace it with some kind of tombstone.
1. Other?

kennsippell commented:
@newtewt Not imminently clear how I'm going to fix this. For the time being, you can force the update dialog by running `medic-conf upload-app-settings` or making a change to the `settings` document in Fauxton.

kennsippell commented:
> To upgrade to 3.5, we have the following requirements:
> 
> 1. Clear the appcache's "master" record for the route /medic/_design/medic/_rewrire/ or else the page will just load from cache and not get the new version
> 1. Trigger the "update now" prompt
> 1. Clear the appCache

I think I've got this working after discussion with @garethbowen.

When given an empty manifest, the appCache wasn't updating `/medic/_design/medic/_rewire/` because that url was yielding a redirect. Per the [AppCache spec](https://www.w3.org/TR/2011/WD-html5-20110525/offline.html): "Redirects are fatal because they are either indicative of a network problem". So I've removed the redirect from the rewrite to root.

So here is how we trigger upgrades in this PR:

1. Changes in `_design/medic-client` trigger the call to `applicationCache.update()`. For now, I am adding a `temp` attribute into the doc but when #5640 this can be removed.
1. The appCache update makes a request to `_rewrite/manifest.appcache`, which I have stubbed to serve an empty manifest. This empty manifest refreshes the appcache `master record' which is `/medic/_design/medic/_rewire/`. I've replaced this with a new `upgrade.html`. (Requirement 1 Achieved)
1. Because the appCache update is successful, the update dialog pops ups (Requirement 2 Acheived)
1. The page now loads and renders `upgrade.html` which clears the AppCache because it requests no manifest. (Requirement 3 Acheived) and then it redirects to root '/'
1. 3.5 + service worker is now loaded. Hurray!

garethbowen commented on ddocs/medic/_attachments/appcache-upgrade.html:
```diff
@@ -0,0 +1,6 @@
+<html>
+  <head>
+    <script type="text/javascript" src="/appcache-upgrade.js"></script>
```
Why not just inline the JS here? I don't think there's any point making a separate request, is there?

kennsippell replied:
The CSP would mean we'd have to maintain a whitelisted hash of the inline script. Not sure which is cleaner.

garethbowen replied:
Of course... I always forget about the CSP.

I wonder if the refresh meta tag would work, eg:

```html
<meta http-equiv="refresh" content="0; url=/" />
```

kennsippell replied:
Ya. Seems to work fine. Tested on mobile + desktop.

garethbowen commented on api/src/routing.js:
```diff
@@ -181,14 +181,17 @@ app.get('/dbinfo', (req, res) => {
+app.get(appPrefix, (req, res, next) => {
+  if (
+    ('deviceID' in req.query) || 
+    req.path.endsWith('/_design/medic/_rewrite/') || // when upgrading from appcache to service-workers
```
I think this is always true. ie:

```js
const appPrefix = pathPrefix + '_design/' + environment.ddoc + '/_rewrite/';
app.get(appPrefix, (req, res, next) => {
  // should only fire when the path is exactly: "/medic/_design/medic/_rewrite/"
});
```

I think it's still ok, it just means we can get rid of the whole `if` block and just `proxy.web` every time.

kennsippell replied:
Done

garethbowen commented on webapp/src/templates/inbox.html:
```diff
@@ -1,6 +1,6 @@
 <!DOCTYPE html>
 
-<html manifest="/empty.manifest">
+<html>
```
:tada:

garethbowen commented on ddocs/medic-client/temp:
```diff
@@ -0,0 +1 @@
+1
```
Leaving a note here to remind whoever merges this to delete this file once this is merged: https://github.com/medic/medic/pull/5640

newtewt commented:
@garethbowen  @kennsippell  this looks good! Seeing the popup and service worker registering and offline working well.  I think after the changes are fixed feel free to close the PR.
